### PR TITLE
Add Qmin and Qmax limits

### PR DESCRIPTION
**Pull request by Saelyos**
Add and optional minimum and maximum limit to the reactive power of flexible parameters #126

### CODE REVIEW

File: roseau/load_flow/models/loads/flexible_parameters.py [hunk 1]
```diff
@@ -477,26 +485,75 @@ def __init__(self, control_p: Control, control_q: Control, projection: Projectio
...
+        self._s_max = value
+        if value <= 0:
+            msg = f"'s_max' must be greater than 0 but {self.s_max:P#~} was provided."
+            logger.error(msg)
+            raise RoseauLoadFlowException(msg=msg, code=RoseauLoadFlowExceptionCode.BAD_FLEXIBLE_PARAMETER_VALUE)
+        if self._q_max is not None and self._q_max > self._s_max:
+            logger.warning("'s_max' has been updated but now 'q_max' is greater than s_max. 'q_max' is set to s_max")
+            self._q_max = self._s_max
+        if self._q_min is not None and self._q_min < -self._s_max:
+            logger.warning("'s_max' has been updated but now 'q_min' is lesser than -s_max. 'q_min' is set to -s_max")
```
Comment (alihamdan) [inline roseau/load_flow/models/loads/flexible_parameters.py:524]: Typo: lesser -> less

File: roseau/load_flow/models/loads/flexible_parameters.py [hunk 2]
```diff
@@ -477,26 +485,75 @@ def __init__(self, control_p: Control, control_q: Control, projection: Projectio
...
+    def q_max(self) -> Q_[float]:
+        """The maximum reactive power of the flexible load (VAr)."""
+        return self._q_max if self._q_max is not None else self._s_max
+
+    @q_max.setter
+    @ureg_wraps(None, (None, "VAr"), strict=False)
+    def q_max(self, value: Optional[float]) -> None:
+        self._q_max = value
+        if value is not None and value > self._s_max:
+            msg = f"'q_max' must be lesser than s_max ({self.s_max:P#~}) but {self.q_max:P#~} was provided."
```
Comment (alihamdan) [inline roseau/load_flow/models/loads/flexible_parameters.py:553]: Same typo

File: roseau/load_flow/models/tests/test_flexible_parameters.py [hunk 3]
```diff
@@ -200,7 +200,29 @@ def test_flexible_parameter():
...
+    with pytest.raises(RoseauLoadFlowException) as e:
+        FlexibleParameter(
+            control_p=Control.constant(),
+            control_q=Control.constant(),
+            projection=Projection(type="euclidean"),
+            s_max=Q_(1e3, "kVA"),
+            q_max=Q_(2e3, "kVAr"),
+        )
+    assert e.value.msg == "'q_max' must be lesser than s_max (1.0 MVA) but 2.0 MVAr was provided."
+    assert e.value.code == RoseauLoadFlowExceptionCode.BAD_FLEXIBLE_PARAMETER_VALUE
```
Comment (alihamdan) [inline roseau/load_flow/models/tests/test_flexible_parameters.py:226]: Maybe add a test for the case where Smax is changed to a value >Qmax or <Qmin?

File: doc/models/Load/FlexibleLoad/FeasibleDomain.md [hunk 4]
```diff
@@ -584,6 +584,96 @@ Notice that when the voltages were too low or too high, the Euclidean projection
 reduction of the produced active power and the (produced and consumed) reactive power to ensure a
 feasible solution. Like before, there is one point per volt and several points overlap.
 
+### $Q^{\min}$ and $Q^{\max}$ limits
+
+It is also possible to define a minimum and maximum reactive power value. In that case, the feasible domain is
```
Comment (alihamdan) [inline doc/models/Load/FlexibleLoad/FeasibleDomain.md:589]:
```suggestion
It is also possible to define a minimum and maximum reactive power values. In that case, the feasible domain is
```

File: doc/models/Load/FlexibleLoad/FeasibleDomain.md [hunk 5]
```diff
@@ -584,6 +584,96 @@ Notice that when the voltages were too low or too high, the Euclidean projection
...
+
+# and when the theoretical power is the following
+power = Q_(-2.5, "kVA")
+
+# Get the resulting flexible powers for the given theoretical power and voltages list.
+auth = ("username", "password")
+res_flexible_powers = fp.compute_powers(auth=auth, voltages=voltages, power=power)
+```
+
+The variable `res_flexible_powers` contains the powers that have been really produced by
```
Comment (alihamdan) [inline doc/models/Load/FlexibleLoad/FeasibleDomain.md:625]:
```suggestion
The variable `res_flexible_powers` contains the powers that have been actually produced by
```
There is another occurrence in this file that I forgot to fix, maybe you can fix it as well.

File: doc/models/Load/FlexibleLoad/FeasibleDomain.md [hunk 6]
```diff
@@ -584,6 +584,96 @@ Notice that when the voltages were too low or too high, the Euclidean projection
...
+the flexible load for the voltages stored in the variable named `voltages`.
+
+Plotting the control curve $Q(U)$ gives:
+
+```{image} /_static/Load/FlexibleLoad/Pconst_QU_Qmin_Qmax_Control_Curve_Example.svg
+:width: 700
+:align: center
+```
+
+Here, again the complete reactive power range is not fully used.
```
Comment (alihamdan) [inline doc/models/Load/FlexibleLoad/FeasibleDomain.md:635]: Also mention the Q constraints. After all this section is about Qmin and Qmax.

**Review by alihamdan — COMMENTED** [review]
Looks good. Left you a few minor comments.

File: roseau/load_flow/models/loads/flexible_parameters.py [hunk 7]
```diff
@@ -840,14 +933,25 @@ def from_dict(cls, data: JsonDict) -> Self:
...
+        )
 
     def to_dict(self, include_geometry: bool = True) -> JsonDict:
         return {
             "control_p": self.control_p.to_dict(),
             "control_q": self.control_q.to_dict(),
             "projection": self.projection.to_dict(),
             "s_max": self._s_max,
+            "q_min": self._q_min,
+            "q_max": self._q_max,
```
Comment (alihamdan) [inline roseau/load_flow/models/loads/flexible_parameters.py:954]: These should only be added if they are not None. This should fix CI

**Review by benoit9126 — CHANGES_REQUESTED** [review]
Great! I wrote some minor comments.

Could you please add a line in the changelog?

**Review by alihamdan — APPROVED** [review]

**Comment by alihamdan** [comment]
> Could you please add a line in the changelog?

@Saelyos you forgot this

**Pull request merged**